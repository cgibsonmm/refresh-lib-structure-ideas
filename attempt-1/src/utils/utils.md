# Utils

This folder will house global used Utils

I.E.

1. commaFormat
