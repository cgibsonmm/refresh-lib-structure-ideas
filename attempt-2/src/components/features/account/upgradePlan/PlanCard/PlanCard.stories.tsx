import { StoryObj, Meta } from '@storybook/react';
import { PlanCard } from './PlanCard';

const meta: Meta<typeof PlanCard> = {
  title: 'Components/UpgradePlan/PlanCard',
  component: PlanCard,
  argTypes: {
    onClick: { action: 'onClick' },
  },
  args: {
    heading: 'Starter',
    limit: 100,
    cost: 29.99,
    renewalPeriod: 1,
  },
};

export default meta;

type Story = StoryObj<typeof PlanCard>;

export const Default: Story = {};

export const SubHeading: Story = {
  args: {
    subheading: 'Starting small and simple',
  },
};

// Hover card variant was disabled due to an issue with Safari, even though the flag can be passed in it has no effect
// export const HoverCard: Story: Story = {
//   args: {
//     hoverCard: true,
//   },
// };

export const CostPerReport: Story = {
  args: {
    costPerReport: 0.3,
  },
};

export const MostPopular: Story = {
  args: {
    mostPopular: true,
  },
};

export const Savings: Story = {
  args: {
    savings: 15,
    costPerReport: 0.3,
  },
};

export const Unlimited: Story = {
  args: {
    limit: -1,
  },
};

export const MultiMonth: Story = {
  args: {
    renewalPeriod: 3,
  },
};

export const Annual: Story = {
  args: {
    renewalPeriod: 12,
  },
};
