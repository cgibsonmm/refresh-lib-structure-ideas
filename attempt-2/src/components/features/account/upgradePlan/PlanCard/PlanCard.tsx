import { Box, Button, Divider, Icons, Text } from '@/theme/index';
import {
  Heading,
  MoLimitLabel,
  MostPopular,
  PlanCardContainer,
  Savings,
  SavingsPerReportContainer,
  SubHeading,
} from './components';

export interface PlanCardProps {
  /** Card heading, usually the plan name. */
  heading: string;
  /** Card subheading, usually the plan description. */
  subheading?: string;
  /** Plan monthly limit, -1 for unlimited plans. */
  limit: number;
  /** Plan cost. */
  cost: number;
  /** Renewal period expressed in months. */
  renewalPeriod: number;
  /**
   * Monthly cost per report for this plan. Use the `costPerReport` util to get this value since there are
   * important rounding considerations to generate the correct value.
   */
  costPerReport?: number;
  /** Flag to indicate if this is the most popular plan for the brand. */
  mostPopular?: boolean;
  /**
   * Savings percentage relative to the current plan comparing the monthly cost per report. Use the `savings` util
   * to get this value since there are important rounding considerations to generate the correct value.
   */
  savings?: number;
  /** Enable the hover card effect. */
  hoverCard?: boolean;
  /** Callback fired when the user clicks the **Upgrade plan** button. */
  onClick: () => void;
}

/**
 * This component is used to display all relevant plan details for a single plan in the upgrade plan flow.
 * It has multiple variants depending on the props passed to it. Some of the variants allow to highlight
 * the most popular plan and/or show the savings percentage relative to the current plan.
 */
export function PlanCard({
  heading,
  subheading,
  limit,
  cost,
  renewalPeriod,
  costPerReport,
  mostPopular,
  savings,
  onClick,
}: PlanCardProps) {
  let cardClasses = 'plan-card';

  // hoverCard css is broken in Safari, so we disable it for now - RS - 2023-05-03
  //if (hoverCard) cardClasses += ' hover-card';

  if (mostPopular) cardClasses += ' most-popular';

  const monthlyPlan = renewalPeriod === 1;
  const unlimitedPlan = limit < 0;
  const hasSavings = !!savings && savings > 0;
  const limitLabel = unlimitedPlan ? 'Unlimited reports' : `${limit} reports`;
  const renewalPeriodLabel = monthlyPlan ? '/mo' : `/${renewalPeriod} mo`;

  return (
    <PlanCardContainer className={cardClasses}>
      {mostPopular && (
        <MostPopular
          label="Most popular"
          size="small"
          icon={<Icons.Star color="inherit" />}
        />
      )}

      {hasSavings ? (
        <Savings label={`${savings}% Savings`} size="small" />
      ) : (
        <Box height={24} />
      )}

      <div>
        <Heading variant="h2">{heading}</Heading>

        {subheading && <SubHeading variant="body2">{subheading}</SubHeading>}
      </div>

      <Text sx={{ fontSize: '30px', lineHeight: '36px' }}>
        <span>{limitLabel}</span>
        <MoLimitLabel component="span">/mo</MoLimitLabel>
      </Text>

      <Divider />

      <div>
        <Text fontSize="18px" lineHeight="22px" color="text.secondary">
          <Text component="span" fontWeight="bold">
            ${cost.toFixed(2)}
          </Text>
          <Text component="span" ml="6px" fontSize="16px" fontWeight="normal">
            {renewalPeriodLabel}
          </Text>
        </Text>

        {!unlimitedPlan && (
          <SavingsPerReportContainer>
            {hasSavings && (
              <Text component="span" fontWeight="bold">
                {savings}% savings,&nbsp;
              </Text>
            )}
            {costPerReport && (
              <span>${costPerReport.toFixed(2)} per report</span>
            )}
          </SavingsPerReportContainer>
        )}
      </div>

      <Button
        variant="contained"
        color="success"
        size="large"
        onClick={onClick}
      >
        Upgrade plan
      </Button>
    </PlanCardContainer>
  );
}
