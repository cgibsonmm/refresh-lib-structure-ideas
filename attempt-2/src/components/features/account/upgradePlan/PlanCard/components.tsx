import { yellow } from '@mui/material/colors';
import { Box, Pill, styled, Text } from '@/theme/index';

export const PlanCardContainer = styled(Box)(({ theme }) => ({
  position: 'relative',
  display: 'flex',
  flexDirection: 'column',
  gap: '20px',
  width: '350px',
  padding: theme.spacing(2.5, 2),
  overflow: 'inherit',
  border: `1px solid ${theme.palette.grey[300]}`,
  borderRadius: '5px',
  backgroundColor: theme.palette.background.paper,

  [theme.breakpoints.up('sm')]: {
    width: 240,
  },

  [theme.breakpoints.up('md')]: {
    width: 275,
    padding: theme.spacing(3),
  },

  '&.hover-card': {
    transition: '0.1s ease',
    transformStyle: 'preserve-3d',

    '::after': {
      content: '""',
      position: 'absolute',
      top: 5,
      right: 0,
      width: '95%',
      height: '100%',
      backgroundColor: theme.palette.grey[200],
      borderRadius: 5,
      transform: 'translateZ(-1px)',
      transition: '0.1s ease',
    },

    ':hover': {
      transform: 'translateY(-1%)',

      '::after': {
        transform: 'translate3d(1%, 1%, -1px) ',
        backgroundColor: theme.palette.grey[300],
      },
    },
  },

  '&.most-popular': {
    border: `2px solid ${theme.palette.primary.main}`,
  },
}));

export const MostPopular = styled(Pill)(({ theme }) => ({
  position: 'absolute',
  top: 0,
  left: '50%',
  paddingInline: theme.spacing(2),
  borderRadius: 5,
  color: theme.palette.common.white,
  backgroundColor: theme.palette.primary.main,
  fontSize: 12,
  fontWeight: 'bold',
  transform: 'translate(-50%, -50%)',
}));

export const Savings = styled(Pill)(({ theme }) => ({
  width: 'fit-content',
  backgroundColor: yellow.A700,
  color: theme.palette.text.primary,
  fontSize: 12,
  fontWeight: 'bold',
}));

export const Heading = styled(Text)(({ theme }) => ({
  margin: 0,
  fontSize: 30,
  lineHeight: '36px',
  color: theme.palette.primary.main,
}));

export const SubHeading = styled(Text)(({ theme }) => ({
  lineHeight: 1.2,
  minHeight: theme.spacing(5),
  color: theme.palette.text.secondary,
}));

export const MoLimitLabel = styled(Text<'span'>)(({ theme }) => ({
  marginLeft: '6px',
  fontSize: '14px',
  fontWeight: 'bold',
  lineHeight: '17px',
  color: theme.palette.text.secondary,
}));

export const SavingsPerReportContainer = styled(Text)(({ theme }) => ({
  fontSize: 14,
  lineHeight: '22px',
  color: theme.palette.primary.main,
}));
