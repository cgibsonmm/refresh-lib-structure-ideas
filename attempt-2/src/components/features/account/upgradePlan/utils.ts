/**
 * Calculates the cost per month for a plan
 *
 * @param cost Cost of the plan
 * @param renewalPeriod Number of months the plan is valid for
 * @returns Cost per month
 */
export function monthlyPlanCost(cost: number, renewalPeriod: number): number {
  return cost / renewalPeriod;
}

/**
 * Calculates the monthly cost per report for a plan
 *
 * @param cost Cost of the plan
 * @param renewalPeriod Number of months the plan is valid for
 * @param limit Monthly report limit of the plan
 * @returns Cost per report rounded to the nearest hundredth
 */
export function costPerReport(
  cost: number,
  renewalPeriod: number,
  limit: number
): number {
  const mthlyPlanCost = Math.ceil(monthlyPlanCost(cost, renewalPeriod));

  // Do not change, the rounding used here is important and was agreed with the product team
  //round to hundredths
  const costPerReport = Math.ceil((mthlyPlanCost / limit) * 100) / 100;

  return costPerReport;
}

/**
 * Calculates the savings based on the monthly cost per report of upgrading from a base plan to a new plan
 *
 * @param baseCost Cost of the base plan
 * @param baseRenewalPeriod Number of months the base plan is valid for
 * @param baseLimit Monthly report limit of the base plan
 * @param upgradeCost Cost of the upgrade plan
 * @param upgradeRenewalPeriod Number of months the new plan is valid for
 * @param upgradeLimit Monthly report limit of the upgrade plan
 * @returns Savings as a percentage
 */
export function savings(
  baseCost: number,
  baseRenewalPeriod: number,
  baseLimit: number,
  upgradeCost: number,
  upgradeRenewalPeriod: number,
  upgradeLimit: number
): number {
  const baseCostPerReport = costPerReport(
    baseCost,
    baseRenewalPeriod,
    baseLimit
  );
  const upgradeCostPerReport = costPerReport(
    upgradeCost,
    upgradeRenewalPeriod,
    upgradeLimit
  );

  // Do not change, the rounding used here is important and was agreed with the product team
  const savings = Math.floor(
    ((baseCostPerReport - upgradeCostPerReport) / baseCostPerReport) * 100
  );

  return savings;
}
