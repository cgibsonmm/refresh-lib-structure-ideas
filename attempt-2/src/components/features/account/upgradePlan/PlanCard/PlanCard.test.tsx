import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { disableReactQueryErrorLogs } from '@/utils/TestUtils';
import { PlanCard, PlanCardProps } from './PlanCard';

disableReactQueryErrorLogs();

const props: PlanCardProps = {
  heading: 'Starter',
  subheading: 'Starting small and simple',
  limit: 100,
  cost: 29.99,
  renewalPeriod: 1,
  costPerReport: 0.33,
  onClick: jest.fn(),
};

test('renders the correct content', () => {
  const { asFragment } = render(<PlanCard {...props} />);

  expect(asFragment()).toMatchSnapshot();
});

test('renders required fields', () => {
  const { container } = render(<PlanCard {...props} />);

  expect(screen.getByRole('heading')).toHaveTextContent('Starter');
  expect(screen.getByText('Starting small and simple')).toBeInTheDocument();
  expect(screen.getByText('100 reports')).toBeInTheDocument();
  expect(screen.getByText('$29.99')).toBeInTheDocument();
  expect(screen.getAllByText('/mo')).toHaveLength(2);
  expect(screen.getByText('$0.33 per report')).toBeInTheDocument();
  expect(screen.getByRole('button')).toHaveTextContent('Upgrade plan');

  expect(container.firstChild).toHaveClass('plan-card');
  expect(container.firstChild).not.toHaveClass('hover-card');
  expect(container.firstChild).not.toHaveClass('most-popular');
  expect(screen.queryByText(/Most popular/)).not.toBeInTheDocument();
  expect(screen.queryByText(/% savings/i)).not.toBeInTheDocument();
});

// hoverCard css is broken in Safari, so we disable it for now - RS - 2023-05-03
// test('hover card', () => {
//   const { container } = render(<PlanCard {...props} hoverCard={true} />);

//   expect(container.firstChild).toHaveClass('hover-card');
// });

test('most popular', () => {
  const { container } = render(<PlanCard {...props} mostPopular={true} />);

  expect(container.firstChild).toHaveClass('most-popular');
  expect(screen.getByText(/Most popular/)).toBeInTheDocument();
});

test('savings', () => {
  render(<PlanCard {...props} savings={12} />);

  expect(screen.getAllByText(/12% savings/i)).toHaveLength(2);
});

test('unlimited', () => {
  render(<PlanCard {...props} limit={-1} />);

  expect(screen.getByText('Unlimited reports')).toBeInTheDocument();
});

test('multi month', () => {
  render(<PlanCard {...props} renewalPeriod={3} />);

  expect(screen.getByText('/mo')).toBeInTheDocument();
  expect(screen.getByText('/3 mo')).toBeInTheDocument();
});

test('decimal positions', () => {
  render(<PlanCard {...props} cost={20} costPerReport={2} />);

  expect(screen.getByText('$20.00')).toBeInTheDocument();
  expect(screen.getByText('$2.00 per report')).toBeInTheDocument();
});

test('selects plan', async () => {
  render(<PlanCard {...props} />);

  const selectPlan = screen.getByRole('button');

  await userEvent.click(selectPlan);

  expect(props.onClick).toHaveBeenCalled();
});
