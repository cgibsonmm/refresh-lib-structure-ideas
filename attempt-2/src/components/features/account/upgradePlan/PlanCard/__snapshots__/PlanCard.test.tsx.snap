// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct content 1`] = `
<DocumentFragment>
  <div
    class="plan-card MuiBox-root css-1274504"
  >
    <div
      class="MuiBox-root css-15p9n5u"
    />
    <div>
      <h2
        class="MuiTypography-root MuiTypography-h2 css-z9vhew-MuiTypography-root"
      >
        Starter
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body2 css-lj764l-MuiTypography-root"
      >
        Starting small and simple
      </p>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-d886p9-MuiTypography-root"
    >
      <span>
        100 reports
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 css-r1ig40-MuiTypography-root"
      >
        /mo
      </span>
    </p>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1gn6kd-MuiTypography-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          $29.99
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 css-mquo0y-MuiTypography-root"
        >
          /mo
        </span>
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-14nz9tk-MuiTypography-root"
      >
        <span>
          $0.33 per report
        </span>
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge css-mkazzp-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Upgrade plan
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
